src/vm-vcpu/src/vcpu/mod.rs

```diff
@@ -240,16 +268,48 @@ impl KvmVcpu {
         self.vcpu_fd.set_lapic(&klapic).map_err(Error::KvmIoctl)
     }
 
+    pub(crate) fn setup_signal_handler() -> Result<()> {
+        extern "C" fn handle_signal(_: c_int, _: *mut siginfo_t, _: *mut c_void) {
+            KvmVcpu::set_local_immediate_exit(1);
+        }
+        #[allow(clippy::identity_op)]
+        register_signal_handler(SIGRTMIN() + 0, handle_signal)
+            .map_err(Error::RegisterSignalHandler)?;
+        Ok(())
+    }
+
+    fn init_tls(&mut self) -> Result<()> {
+        Self::TLS_VCPU_PTR.with(|vcpu| {
+            if vcpu.borrow().is_none() {
+                *vcpu.borrow_mut() = Some(self as *const KvmVcpu);
+                Ok(())
+            } else {
+                Err(Error::TlsInitialized)
+            }
+        })?;
+        Ok(())
+    }
+
+    fn set_local_immediate_exit(value: u8) {
+        Self::TLS_VCPU_PTR.with(|v| {
+            if let Some(vcpu) = *v.borrow() {
+                unsafe {
+                    let vcpu_ref = &*vcpu;
+                    vcpu_ref.vcpu_fd.set_kvm_immediate_exit(value);
+                };
+            }
+        });
+    }
+
     /// vCPU emulation loop.
     #[allow(clippy::if_same_then_else)]
     pub fn run(&mut self, instruction_pointer: GuestAddress) -> Result<()> {
-        if !self.running {
-            self.configure_regs(instruction_pointer)?;
-            self.running = true;
-        }
-        self.run_barrier.wait();
+        self.configure_regs(instruction_pointer)?;
+        self.init_tls()?;
 
+        self.run_barrier.wait();
         loop {
+            let mut interrupted_by_signal = false;
```

Comment:
Not really important, but I think renaming `interrupted_by_signal` to just `interrupted` sounds better.